File: LarkCardKit/Builders/InputBuilder.cs
using LarkCardKit.Enums;
using LarkCardKit.Models.Behaviors;
using LarkCardKit.Models.Elements;
using LarkCardKit.Templates;

namespace LarkCardKit.Builders;

/// <summary>
/// 输入框构建器，用于构建文本输入组件
/// </summary>
/// <remarks>
/// 输入框支持多种输入类型：
/// <list type="bullet">
///   <item><description>Text - 单行文本</description></item>
///   <item><description>MultilineText - 多行文本</description></item>
///   <item><description>Password - 密码输入</description></item>
/// </list>
/// </remarks>
/// <example>
/// 以下示例演示如何使用 InputBuilder：
/// <code>
/// var input = new InputBuilder()
///     .Name("username")
///     .Placeholder("请输入用户名")
///     .Required()
///     .Build();
/// </code>
/// </example>
public class InputBuilder
{
    private readonly Input _input = new();
    
    /// <summary>
    /// 设置输入框的唯一标识名称（表单容器中必填）
    /// </summary>
    /// <param name="name">输入框名称</param>
    /// <returns>当前构建器实例（支持链式调用）</returns>
    public InputBuilder Name(string name)
    {
        _input.Name = name;
        return this;
    }
    
    /// <summary>
    /// 设置是否必填
    /// </summary>
    /// <param name="required">是否必填，默认为 true</param>
    /// <returns>当前构建器实例（支持链式调用）</returns>
    public InputBuilder Required(bool required = true)
    {
        _input.Required = required;
        return this;
    }
    
    /// <summary>
    /// 设置占位提示文本
    /// </summary>
    /// <param name="placeholder">占位文本内容</param>
    /// <returns>当前构建器实例（支持链式调用）</returns>
    public InputBuilder Placeholder(string placeholder)
    {
        _input.Placeholder = new PlainText { Content = placeholder };
        return this;
    }
    
    /// <summary>
    /// 设置默认值
    /// </summary>
    /// <param name="value">默认值内容</param>
    /// <returns>当前构建器实例（支持链式调用）</returns>
    public InputBuilder DefaultValue(string value)
    {
        _input.DefaultValue = value;
        return this;
    }
    
    /// <summary>
    /// 设置输入类型
    /// </summary>
    /// <param name="type">
    /// 输入类型
    /// <para>Text - 单行文本</para>
    /// <para>MultilineText - 多行文本</para>
    /// <para>Password - 密码输入</para>
    /// </param>
    /// <returns>当前构建器实例（支持链式调用）</returns>
    public InputBuilder Type(InputType type)
    {
        _input.InputType = type switch
        {
            InputType.Text => "text",
            InputType.MultilineText => "multiline_text",
            InputType.Password => "password",
            _ => "text"
        };
        return this;
    }
    
    /// <summary>
    /// 设置文本标签
    /// </summary>
    /// <param name="label">标签文本内容</param>
    /// <returns>当前构建器实例（支持链式调用）</returns>
    public InputBuilder Label(string label)
    {
        _input.Label = new PlainText { Content = label };
        return this;
    }
    
    /// <summary>
    /// 设置标签位置
    /// </summary>
    /// <param name="position">标签位置：top（上方）或 left（左侧）</param>
    /// <returns>当前构建器实例（支持链式调用）</returns>
    public InputBuilder LabelPosition(string position)
    {
        _input.LabelPosition = position;
        return this;
    }
    
    /// <summary>
    /// 设置最大输入长度
    /// </summary>
    /// <param name="length">最大长度，取值范围 1-1000</param>
    /// <returns>当前构建器实例（支持链式调用）</returns>
    public InputBuilder MaxLength(int length)
    {
        _input.MaxLength = length;
        return this;
    }
    
    /// <summary>
    /// 设置多行文本的默认行数
    /// </summary>
    /// <param name="rows">行数</param>
    /// <returns>当前构建器实例（支持链式调用）</returns>
    public InputBuilder Rows(int rows)
    {
        _input.Rows = rows;
        return this;
    }
    
    /// <summary>
    /// 设置是否自适应高度
    /// </summary>
    /// <param name="resize">是否自适应，默认为 true</param>
    /// <returns>当前构建器实例（支持链式调用）</returns>
    public InputBuilder AutoResize(bool resize = true)
    {
        _input.AutoResize = resize;
        return this;
    }
    
    /// <summary>
    /// 设置多行文本的最大行数
    /// </summary>
    /// <param name="rows">最大行数</param>
    /// <returns>当前构建器实例（支持链式调用）</returns>
    public InputBuilder MaxRows(int rows)
    {
        _input.MaxRows = rows;
        return this;
    }
    
    /// <summary>
    /// 设置输入框宽度
    /// </summary>
    /// <param name="width">宽度值，如 "default", "fill", "200px"</param>
    /// <returns>当前构建器实例（支持链式调用）</returns>
    public InputBuilder Width(string width)
    {
        _input.Width = width;
        return this;
    }
    
    /// <summary>
    /// 设置元素唯一标识
    /// </summary>
    /// <param name="id">元素 ID</param>
    /// <returns>当前构建器实例（支持链式调用）</returns>
    public InputBuilder ElementId(string id)
    {
        _input.ElementId = id;
        return this;
    }
    
    /// <summary>
    /// 设置外边距
    /// </summary>
    /// <param name="margin">外边距值，如 "4px", "8px 12px"</param>
    /// <returns>当前构建器实例（支持链式调用）</returns>
    public InputBuilder Margin(string margin)
    {
        _input.Margin = margin;
        return this;
    }
    
    /// <summary>
    /// 设置是否禁用
    /// </summary>
    /// <param name="disabled">是否禁用，默认为 true</param>
    /// <returns>当前构建器实例（支持链式调用）</returns>
    public InputBuilder Disabled(bool disabled = true)
    {
        _input.Disabled = disabled;
        return this;
    }

    /// <summary>
    /// 设置是否禁用（使用模板值）
    /// </summary>
    /// <param name="value">模板值对象</param>
    /// <returns>当前构建器实例（支持链式调用）</returns>
    /// <remarks>
    /// 使用此方法时，禁用状态将在调用 <see cref="CardBuilder.SetParameter"/> 时动态填充
    /// </remarks>
    public InputBuilder Disabled(TemplateValue value)
    {
        var placeholder = value.DefaultValue != null
            ? $"${{{value.Key}:{value.DefaultValue}}}"
            : $"${{{value.Key}}}";
        _input.Disabled = placeholder;
        return this;
    }
    
    /// <summary>
    /// 设置禁用提示文本
    /// </summary>
    /// <param name="tips">禁用提示内容</param>
    /// <returns>当前构建器实例（支持链式调用）</returns>
    public InputBuilder DisabledTips(string tips)
    {
        _input.DisabledTips = new PlainText { Content = tips };
        return this;
    }
    
    /// <summary>
    /// 设置输入变更时的回调数据
    /// </summary>
    /// <param name="callbackData">回调数据对象</param>
    /// <returns>当前构建器实例（支持链式调用）</returns>
    public InputBuilder OnChange(object callbackData)
    {
        _input.Behaviors ??= new List<object>();
        _input.Behaviors.Add(new CallbackBehavior { Value = callbackData });
        return this;
    }
    
    /// <summary>
    /// 构建输入框对象
    /// </summary>
    /// <returns>完整的 <see cref="Input"/> 对象</returns>
    public Input Build() => _input;
}

File: Python/lark_card_kit/builders/input_builder.py
"""输入框构建器，提供流畅 API，支持完美的类型提示。

本模块只支持两种构建方式：
1. 上下文管理器（推荐用于嵌套结构）
2. 方法链 + done()（推荐用于简单结构）
"""

from __future__ import annotations

from typing import TYPE_CHECKING, Any, Self

from ..enums import InputType
from ..models.elements import Input, PlainText

if TYPE_CHECKING:
    from .card_body_builder import CardBodyBuilder
    from .form_builder import FormBuilder


class InputBuilder:
    """输入框构建器，提供流畅 API。

    示例:
        >>> # 方式一：上下文管理器
        >>> with builder.body().input() as i:
        ...     i.name("username").placeholder("Enter username").required()
        >>>
        >>> # 方式二：方法链 + done()
        >>> builder.body().input().name("username").placeholder("Enter").done()
    """

    def __init__(self, parent: CardBodyBuilder | FormBuilder | None = None) -> None:
        """初始化输入框构建器。

        参数:
            parent: 父构建器，用于嵌套构建。
        """
        self._input = Input()
        self._parent = parent

    def __enter__(self) -> Self:
        """进入上下文管理器。

        返回:
            自身，供上下文内使用。
        """
        return self

    def __exit__(self, *args: Any) -> None:
        """退出上下文管理器，将输入框添加到父构建器。"""
        if self._parent is not None:
            self._parent.add_element(self._input)

    def name(self, name: str) -> Self:
        """设置输入框名称（用于表单提交）。

        参数:
            name: 输入框名称。

        返回:
            自身，用于链式调用。
        """
        self._input.name = name
        return self

    def placeholder(self, placeholder: str) -> Self:
        """设置占位文本。

        参数:
            placeholder: 占位文本。

        返回:
            自身，用于链式调用。
        """
        self._input.placeholder = PlainText(content=placeholder)
        return self

    def default_value(self, value: str) -> Self:
        """设置默认值。

        参数:
            value: 默认值。

        返回:
            自身，用于链式调用。
        """
        self._input.default_value = value
        return self

    def input_type(self, input_type: InputType | str) -> Self:
        """设置输入类型。

        参数:
            input_type: 输入类型（text, password, number, email）。

        返回:
            自身，用于链式调用。
        """
        self._input.input_type = input_type.value if isinstance(input_type, InputType) else input_type
        return self

    def label(self, label: str) -> Self:
        """设置标签。

        参数:
            label: 标签文本。

        返回:
            自身，用于链式调用。
        """
        self._input.label = PlainText(content=label)
        return self

    def label_position(self, position: str) -> Self:
        """设置标签位置。

        参数:
            position: 标签位置（如 'left', 'top'）。

        返回:
            自身，用于链式调用。
        """
        self._input.label_position = position
        return self

    def required(self, required: bool = True) -> Self:
        """设置是否必填。

        参数:
            required: 是否必填。

        返回:
            自身，用于链式调用。
        """
        self._input.required = required
        return self

    def max_length(self, max_length: int) -> Self:
        """设置最大长度。

        参数:
            max_length: 最大长度。

        返回:
            自身，用于链式调用。
        """
        self._input.max_length = max_length
        return self

    def rows(self, rows: int) -> Self:
        """设置多行输入的行数。

        参数:
            rows: 行数。

        返回:
            自身，用于链式调用。
        """
        self._input.rows = rows
        return self

    def auto_resize(self, auto_resize: bool = True) -> Self:
        """设置是否自动调整大小。

        参数:
            auto_resize: 是否自动调整大小。

        返回:
            自身，用于链式调用。
        """
        self._input.auto_resize = auto_resize
        return self

    def max_rows(self, max_rows: int) -> Self:
        """设置最大行数。

        参数:
            max_rows: 最大行数。

        返回:
            自身，用于链式调用。
        """
        self._input.max_rows = max_rows
        return self

    def width(self, width: str) -> Self:
        """设置输入框宽度。

        参数:
            width: 输入框宽度。

        返回:
            自身，用于链式调用。
        """
        self._input.width = width
        return self

    def disabled(self, disabled: bool = True) -> Self:
        """设置是否禁用。

        参数:
            disabled: 是否禁用。

        返回:
            自身，用于链式调用。
        """
        self._input.disabled = disabled
        return self

    def disabled_tips(self, tips: str) -> Self:
        """设置禁用提示。

        参数:
            tips: 禁用时显示的提示。

        返回:
            自身，用于链式调用。
        """
        self._input.disabled_tips = PlainText(content=tips)
        return self

    def element_id(self, element_id: str) -> Self:
        """设置元素 ID。

        参数:
            element_id: 元素 ID。

        返回:
            自身，用于链式调用。
        """
        self._input.element_id = element_id
        return self

    def done(self) -> CardBodyBuilder | FormBuilder:
        """完成输入框构建，返回父构建器。

        返回:
            父构建器实例。

        异常:
            RuntimeError: 如果没有使用父构建器。
        """
        if self._parent is None:
            raise RuntimeError("done() 只能在嵌套构建时调用")
        self._parent.add_element(self._input)
        return self._parent

    def build(self) -> Input:
        """构建输入框元素。

        返回:
            构建好的 Input 实例。
        """
        return self._input